core/src/main/scala/org/bitcoins/core/crypto/TransactionSignatureSerializer.scala

```diff
@@ -253,6 +254,233 @@ sealed abstract class TransactionSignatureSerializer {
+        val isNotAnyoneCanPay = !HashType.isAnyoneCanPay(hashType)
+        val isNotSigHashSingle = !(HashType.isSigHashSingle(hashType.num))
+        val isNotSigHashNone = !(HashType.isSigHashNone(hashType.num))
+        val inputIndexInt = idx
+        val emptyHash = DoubleSha256Digest.empty
+
+        val outPointHash: ByteVector = if (isNotAnyoneCanPay) {
+          val prevOuts = spendingTransaction.inputs.map(_.previousOutput)
+          val bytes: ByteVector = BytesUtil.toByteVector(prevOuts)
+          CryptoUtil.doubleSHA256(bytes).bytes
+        } else emptyHash.bytes
+
+        val sequenceHash: ByteVector =
+          if (isNotAnyoneCanPay && isNotSigHashNone && isNotSigHashSingle) {
+            val sequences = spendingTransaction.inputs.map(_.sequence)
+            val littleEndianSeq =
+              sequences.foldLeft(ByteVector.empty)(_ ++ _.bytes.reverse)
+            CryptoUtil.doubleSHA256(littleEndianSeq).bytes
+          } else emptyHash.bytes
+
+        val outputHash: ByteVector =
+          if (isNotSigHashSingle && isNotSigHashNone) {
+            val outputs = spendingTransaction.outputs
+            val bytes = BytesUtil.toByteVector(outputs)
+            CryptoUtil.doubleSHA256(bytes).bytes
+          } else if (
+            HashType.isSigHashSingle(hashType.num) &&
+            idx < spendingTransaction.outputs.size
+          ) {
+            val output = spendingTransaction.outputs(inputIndexInt)
+            val bytes = CryptoUtil
+              .doubleSHA256(RawTransactionOutputParser.write(output))
+              .bytes
+            bytes
+          } else emptyHash.bytes
+
+        val scriptBytes = BytesUtil.toByteVector(script)
+
+        val i = spendingTransaction.inputs(inputIndexInt)
+
+        val serializationForSig: ByteVector =
+          spendingTransaction.version.bytes.reverse ++ outPointHash ++ sequenceHash ++
+            i.previousOutput.bytes ++ CompactSizeUInt.calc(scriptBytes).bytes ++
+            scriptBytes ++ amount.bytes ++ i.sequence.bytes.reverse ++
+            outputHash ++ spendingTransaction.lockTime.bytes.reverse ++ hashType.num.bytes.reverse
+        logger.debug(
+          "Serialization for signature for WitnessV0Sig: " + BytesUtil
+            .encodeHex(serializationForSig))
+        serializationForSig
+    }
+  }
+
+  /**
+    * Hashes a [[org.bitcoins.core.wallet.utxo.InputSigningInfo InputSigningInfo]] to give the value that needs to be signed
+    * by a [[org.bitcoins.crypto.Sign Sign]] to
+    * produce a valid [[org.bitcoins.crypto.ECDigitalSignature ECDigitalSignature]] for a transaction
+    */
+  def hashForSignature(
+      spendingTransaction: Transaction,
+      signingInfo: InputSigningInfo[InputInfo],
+      hashType: HashType): DoubleSha256Digest = {
+    val inputIndexOpt = spendingTransaction.inputs.zipWithIndex
+      .find(_._1.previousOutput == signingInfo.outPoint)
+      .map(_._2)
+
+    if (inputIndexOpt.isEmpty) {
```

Comment:
I would prefer to move this into `serializeForSignature()` as it can be used indepdently of `hashForSignature()`. Basically this check should be moved to the lowest common denominator for signing things (serialization for signature, not hashing).

Reply:
This needs to stay here unless we know the preimage to `errorHash`, however, I make it so `serializeForSignature` will throw an error if the same parameters are given there.